File: native-android/app/src/main/kotlin/com/crossdashboard/app/data/network/NextcloudLoginFlow.kt
package com.crossdashboard.app.data.network

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Nextcloud Login Flow v2 (browser-based).
 *
 * Flow:
 * 1. POST to /index.php/login/v2 → get loginUrl + pollToken + pollEndpoint
 * 2. Open loginUrl in a Custom Tab — user approves in browser
 * 3. Poll pollEndpoint with pollToken until credentials arrive (max 5 min)
 * 4. Returns (serverUrl, loginName, appPassword) to store in SecureStore
 */
@Singleton
class NextcloudLoginFlow @Inject constructor(
    private val httpClient: OkHttpClient,
) {
    private val json = Json { ignoreUnknownKeys = true }

    data class FlowInit(
        val loginUrl: String,
        val pollEndpoint: String,
        val pollToken: String,
    )

    data class LoginCredentials(
        val serverUrl: String,
        val loginName: String,
        val appPassword: String,
    )

    suspend fun initiate(serverUrl: String): Result<FlowInit> = withContext(Dispatchers.IO) {
        try {
            val url = "${serverUrl.trimEnd('/')}/index.php/login/v2"
            val body = "".toRequestBody("application/x-www-form-urlencoded".toMediaType())
            val response = httpClient.newCall(
                Request.Builder().url(url).post(body).build()
            ).execute()
            if (!response.isSuccessful) return@withContext Result.failure(Exception("HTTP ${response.code}"))
            val dto = json.decodeFromString<LoginFlowInitDto>(response.body!!.string())
            Result.success(FlowInit(dto.login, dto.poll.endpoint, dto.poll.token))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun poll(
        pollEndpoint: String,
        pollToken: String,
        timeoutMs: Long = 5 * 60_000L,
    ): Result<LoginCredentials> = withContext(Dispatchers.IO) {
        try {
            withTimeout(timeoutMs) {
                val body = "token=$pollToken".toRequestBody("application/x-www-form-urlencoded".toMediaType())
                while (true) {
                    delay(2000)
                    val response = httpClient.newCall(
                        Request.Builder().url(pollEndpoint).post(body).build()
                    ).execute()
                    if (response.code == 200) {
                        val dto = json.decodeFromString<LoginCredentialsDto>(response.body!!.string())
                        return@withTimeout Result.success(
                            LoginCredentials(dto.server, dto.loginName, dto.appPassword)
                        )
                    }
                    // 404 = not yet approved; keep polling
                }
                @Suppress("UNREACHABLE_CODE")
                Result.failure(Exception("Timeout"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    @Serializable
    private data class LoginFlowInitDto(val poll: PollDto, val login: String)

    @Serializable
    private data class PollDto(val token: String, val endpoint: String)

    @Serializable
    private data class LoginCredentialsDto(
        val server: String,
        val loginName: String,
        val appPassword: String,
    )
}

File: native-macos/CrossDashboard/Data/Network/NextcloudLoginFlow.swift
import Foundation
import CrossDashboardKit

/// Nextcloud Login Flow v2 (browser-based).
///
/// Flow:
/// 1. POST to /index.php/login/v2 → get loginUrl + pollToken + pollEndpoint
/// 2. Open loginUrl in system browser (NSWorkspace.open) — user approves
/// 3. Poll pollEndpoint with pollToken every 2s for up to 5 min
/// 4. Returns (serverUrl, loginName, appPassword) to store in KeychainStore
///
/// Direct Swift port of NextcloudLoginFlow.kt.
final class NextcloudLoginFlow: Sendable {

    private let session: URLSession

    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 15
        self.session = URLSession(configuration: config)
    }

    // ─── Data types ───────────────────────────────────────────────────────────

    struct FlowInit: Sendable {
        let loginUrl: String
        let pollEndpoint: String
        let pollToken: String
    }

    struct LoginCredentials: Sendable {
        let serverUrl: String
        let loginName: String
        let appPassword: String
    }

    // ─── Step 1: initiate ─────────────────────────────────────────────────────

    func initiate(serverUrl: String) async -> Result<FlowInit, Error> {
        let urlStr = "\(serverUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")))/index.php/login/v2"
        guard let url = URL(string: urlStr) else {
            return .failure(LoginFlowError.invalidUrl(urlStr))
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = Data()

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                return .failure(LoginFlowError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0))
            }
            let dto = try JSONDecoder().decode(LoginFlowInitDto.self, from: data)
            return .success(FlowInit(
                loginUrl:     dto.login,
                pollEndpoint: dto.poll.endpoint,
                pollToken:    dto.poll.token
            ))
        } catch {
            return .failure(error)
        }
    }

    // ─── Step 3: poll ─────────────────────────────────────────────────────────

    func poll(
        pollEndpoint: String,
        pollToken: String,
        timeoutSeconds: TimeInterval = 5 * 60
    ) async -> Result<LoginCredentials, Error> {
        guard let url = URL(string: pollEndpoint) else {
            return .failure(LoginFlowError.invalidUrl(pollEndpoint))
        }
        let deadline = Date().addingTimeInterval(timeoutSeconds)

        while Date() < deadline {
            try? await Task.sleep(for: .seconds(2))

            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
            request.httpBody = "token=\(pollToken)".data(using: .utf8)

            do {
                let (data, response) = try await session.data(for: request)
                if let http = response as? HTTPURLResponse, http.statusCode == 200 {
                    let dto = try JSONDecoder().decode(LoginCredentialsDto.self, from: data)
                    return .success(LoginCredentials(
                        serverUrl:   dto.server,
                        loginName:   dto.loginName,
                        appPassword: dto.appPassword
                    ))
                }
                // 404 = not yet approved; keep polling
            } catch {
                return .failure(error)
            }
        }
        return .failure(LoginFlowError.timeout)
    }

    // ─── Private DTOs ─────────────────────────────────────────────────────────

    private struct LoginFlowInitDto: Decodable {
        let poll: PollDto
        let login: String
    }

    private struct PollDto: Decodable {
        let token: String
        let endpoint: String
    }

    private struct LoginCredentialsDto: Decodable {
        let server: String
        let loginName: String
        let appPassword: String
    }
}

// ─── Errors ───────────────────────────────────────────────────────────────────

enum LoginFlowError: LocalizedError {
    case invalidUrl(String)
    case httpError(Int)
    case timeout

    var errorDescription: String? {
        switch self {
        case .invalidUrl(let u): return "Invalid URL: \(u)"
        case .httpError(let c):  return "HTTP \(c)"
        case .timeout:           return "Login flow timed out after 5 minutes."
        }
    }
}
